instrumentation/reactor/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/HttpResponseReceiverInstrumenter.java

```diff
@@ -58,151 +55,105 @@ public static HttpClient.ResponseReceiver<?> instrument(HttpClient.ResponseRecei
 
-    PropagateContext(ContextHolder contextHolder) {
-      this.contextHolder = contextHolder;
+    StartOperation(InstrumentationContexts instrumentationContexts) {
+      this.instrumentationContexts = instrumentationContexts;
     }
 
     @Override
-    public void accept(HttpClientRequest httpClientRequest, Connection connection) {
-      Context context = contextHolder.context;
-      if (context != null) {
-        GlobalOpenTelemetry.getPropagators()
-            .getTextMapPropagator()
-            .inject(context, httpClientRequest, HttpClientRequestHeadersSetter.INSTANCE);
-      }
+    public void accept(HttpClientRequest request, Connection connection) {
+      Context context = instrumentationContexts.startClientSpan(request);
 
       // also propagate the context to the underlying netty instrumentation
       // if this span was suppressed and context is null, propagate parentContext - this will allow
       // netty spans to be suppressed too
-      Context nettyParentContext = context == null ? contextHolder.parentContext : context;
+      Context nettyParentContext =
+          context == null ? instrumentationContexts.getParentContext() : context;
       NettyClientTelemetry.setChannelContext(connection.channel(), nettyParentContext);
     }
   }
 
   static final class EndOperationWithRequestError
       implements BiConsumer<HttpClientRequest, Throwable> {
 
-    private final ContextHolder contextHolder;
-    private final HttpClientConfig config;
+    private final InstrumentationContexts instrumentationContexts;
 
-    EndOperationWithRequestError(ContextHolder contextHolder, HttpClientConfig config) {
-      this.contextHolder = contextHolder;
-      this.config = config;
+    EndOperationWithRequestError(InstrumentationContexts instrumentationContexts) {
+      this.instrumentationContexts = instrumentationContexts;
     }
 
     @Override
-    public void accept(HttpClientRequest httpClientRequest, Throwable error) {
-      Context context = contextHolder.getAndRemoveContext();
-      if (context == null) {
-        return;
+    public void accept(HttpClientRequest request, Throwable error) {
+      instrumentationContexts.endClientSpan(null, error);
+
+      if (HttpClientResend.get(instrumentationContexts.getParentContext()) == 0) {
+        // TODO: emit connection error span
+
+        // FIXME: this branch requires lots of changes around the NettyConnectionInstrumenter
+        // currently it also creates that connection error span (when the connection telemetry is
+        // turned off), but without HTTP semantics - it does not have access to any HTTP information
+        // after all
+        // it should be possible to completely disable it, and just start and end the span here
+        // this requires lots of refactoring and pretty uninteresting changes in the netty code, so
+        // I'll do that in a separate PR - for better readability
       }
```

Comment:
Here's where the connection span will be added -- similar to the `ConnectionErrorSpanInterceptor` in the OkHttp instrumentation, it's only emitted on error and when there were no HTTP requests made.